### DISCUSSION TITLE
Way to generate a static image of a process tree programmatically?

### DISCUSSION THREAD

**Discussion by nbareil** [microsoft/msticpy#541] [Ideas]
👋 

Do you have an idea or experience to share on how to generate a static image of a [process tree](https://msticpy.readthedocs.io/en/latest/visualization/ProcessTree.html) programmatically (in Python)?

My end goal is to post the picture directly into our ticketing system.

I saw there were some level of headless support in https://github.com/bokeh/bokeh/issues/538 and especially https://github.com/bokeh/bokeh/issues/6594#issuecomment-346178496 but I am curious to know if anybody already performed a similar thing or if they tried another way?

**Comment by ianhelle** [comment]
This is possible but we don;t have any built-in support for it.
If you check out the bokeh documentation you can use plot.save_* functions to save as png or svg (you need to install addiitional libraries for the PNG export). See https://docs.bokeh.org/en/2.4.1/docs/user_guide/export.html#:~:text=Bokeh%20can%20generate%20RGBA-format%20Portable%20Network%20Graphics%20%28PNG%29,have%20the%20same%20dimensions%20as%20the%20source%20layout.

Most of our plot functions return either the bokeh plot or a Column layout. 
You can check the docs/code to see which is which or check the type() of the return value.
If it's a plot object, you can just invoke the method directly on the returned object.
If it's a column object you'd need to extract the plot object - I think this is just 
layout = df.mp_plot.process_tree()
layout[0].save_png('file.png")

You might need to experiment a bit to see which one works with which function.

**Reply by simsor** [reply]
This kinda works (it's using [bokeh's `save_png`](https://docs.bokeh.org/en/2.4.1/docs/user_guide/export.html#example-usage)):
```python
import pandas as pd
from bokeh.io import export_png
from msticpy.nbtools import process_tree as ptree

proc_df = pd.read_csv("proc_df.csv")
p_tree_win = ptree.build_process_tree(proc_df, debug=True)
fig = ptree.plot_process_tree(p_tree_win)
export_png(fig[0], filename="test.png")
```

Only issue is that there doesn't seem to be the whole tree, we can only see the limited window shown in a Jupyter notebook (which you're supposed to scroll in). I don't know if it's possible to ask for the full size image.

**Reply by ianhelle** [reply]
One possibility might be to wrap the export in an API that re-draws the entire tree (in the Bokeh JS model, not on the screen) before rendering as SVG/PNG. Not sure if this is doable.
You can manually do it by setting the plot height in the call to process tree to something huge and then rendering.
